File: permission/src/main/java/dev/eastar/permission/PermissionChecker.java
/*
 * Copyright 2016 copyright eastar Jeong
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dev.eastar.permission;

import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.provider.Settings;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v7.app.AlertDialog;
import android.view.WindowManager;

import java.util.List;

public class PermissionChecker extends android.support.v7.app.AppCompatActivity {

    private Context mContext;
    private Activity mActivity;

    private static final int REQ_REQUEST = 10;
    private static final int REQ_SETTING = 20;

    public interface EXTRA {
        String PERMISSIONS = "PERMISSIONS";

        String REQUEST_MESSAGE = "REQUEST_MESSAGE";
        String REQUEST_POSITIVE_BUTTON_TEXT = "REQUEST_POSITIVE_BUTTON_TEXT";
        String REQUEST_NEGATIVE_BUTTON_TEXT = "REQUEST_NEGATIVE_BUTTON_TEXT";

        String DENY_MESSAGE = "DENY_MESSAGE";
        String DENY_POSITIVE_BUTTON_TEXT = "DENY_POSITIVE_BUTTON_TEXT";
        String DENY_NEGATIVE_BUTTON_TEXT = "DENY_NEGATIVE_BUTTON_TEXT";
    }

    private List<String> mRequestedPermissions;//최초 물어본 권한

    private CharSequence mRequestMessage;

    private CharSequence mDenyMessage;

    private CharSequence mRequestPositiveButtonText;
    private CharSequence mRequestNegativeButtonText;
    private CharSequence mDenyPositiveButtonText;
    private CharSequence mDenyNegativeButtonText;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mContext = this;
        mActivity = this;
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE);
        parseExtra();
        load();
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        setIntent(intent);
        parseExtra();
        load();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (mDlg != null && mDlg.isShowing())
            mDlg.cancel();
    }

    private void parseExtra() {
        final Intent intent = getIntent();
        mRequestedPermissions = intent.getStringArrayListExtra(EXTRA.PERMISSIONS);

        mRequestMessage = intent.getCharSequenceExtra(EXTRA.REQUEST_MESSAGE);
        mRequestPositiveButtonText = intent.getCharSequenceExtra(EXTRA.REQUEST_POSITIVE_BUTTON_TEXT);
        mRequestNegativeButtonText = intent.getCharSequenceExtra(EXTRA.REQUEST_NEGATIVE_BUTTON_TEXT);

        mDenyMessage = intent.getCharSequenceExtra(EXTRA.DENY_MESSAGE);
        mDenyPositiveButtonText = intent.getCharSequenceExtra(EXTRA.DENY_POSITIVE_BUTTON_TEXT);
        mDenyNegativeButtonText = intent.getCharSequenceExtra(EXTRA.DENY_NEGATIVE_BUTTON_TEXT);

    }

    private AlertDialog mDlg;

    private void load() {
        final List<String> deniedPermissions = PermissionRequest.getDeniedPermissions(mContext, mRequestedPermissions);

        if (deniedPermissions.size() <= 0)
            throw new UnsupportedOperationException("!!deniedPermissions <= 0");
//        android.util.Log.e("PERMISSIONS", "묻어봄" + deniedPermissions);
        requestPermissions(deniedPermissions);
    }

    private void requestPermissions(@NonNull final List<String> deniedPermissions) {
        final CharSequence message = mRequestMessage;
        final String[] permissions = deniedPermissions.toArray(new String[deniedPermissions.size()]);

//        android.util.Log.e("PERMISSIONS", "묻어봄" + Arrays.toString(permissions));

        if (message == null || message.length() <= 0) {
            ActivityCompat.requestPermissions(mActivity, permissions, REQ_REQUEST);
            return;
        }

        final Context context = mContext;
        if (mRequestPositiveButtonText == null || mRequestPositiveButtonText.length() <= 0)
            mRequestPositiveButtonText = context.getString(R.string.permission_confirm);
        if (mRequestNegativeButtonText == null || mRequestNegativeButtonText.length() <= 0)
            mRequestNegativeButtonText = context.getString(R.string.permission_close);
        mDlg = new AlertDialog.Builder(mContext)//
                .setMessage(message)//
                .setOnCancelListener(dialog -> fireDenied())//
                .setPositiveButton(mRequestPositiveButtonText, (dialogInterface, i) -> ActivityCompat.requestPermissions(mActivity, deniedPermissions.toArray(new String[deniedPermissions.size()]), REQ_REQUEST))
                .setNegativeButton(mRequestNegativeButtonText, (dialogInterface, i) -> fireDenied())//
                .show();
        mDlg.setCanceledOnTouchOutside(false);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
//        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == REQ_REQUEST) {
            final List<String> deniedPermissions = PermissionRequest.getDeniedPermissions(mContext, mRequestedPermissions);
            if (deniedPermissions.size() > 0)
                denyPermissions();
            else
                fireGranted();
        }
    }

    public void denyPermissions() {
        final Context context = mContext;
        final CharSequence message = mDenyMessage;

        if (message == null || message.length() <= 0) {
            fireDenied();
            return;
        }

        if (mDenyPositiveButtonText == null || mDenyPositiveButtonText.length() <= 0)
            mDenyPositiveButtonText = context.getString(R.string.permission_close);
        if (mDenyNegativeButtonText == null || mDenyNegativeButtonText.length() <= 0)
            mDenyNegativeButtonText = context.getString(R.string.permission_setting);

        mDlg = new AlertDialog.Builder(context)//
                .setMessage(message)//
                .setOnCancelListener(dialog -> fireDenied())//
                .setPositiveButton(mDenyPositiveButtonText, (dialogInterface, i) -> fireDenied())//
                .setNegativeButton(mDenyNegativeButtonText, (dialogInterface, i) -> {
                    try {
                        Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS).setData(Uri.parse("package:" + context.getPackageName()));
                        startActivityForResult(intent, REQ_SETTING);
                    } catch (ActivityNotFoundException e) {
                        Intent intent = new Intent(Settings.ACTION_MANAGE_APPLICATIONS_SETTINGS);
                        startActivityForResult(intent, REQ_SETTING);
                    }
                })//
                .show();
        mDlg.setCanceledOnTouchOutside(false);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQ_SETTING) {
            final List<String> deniedPermissions = PermissionRequest.getDeniedPermissions(mContext, mRequestedPermissions);
            if (deniedPermissions.size() > 0)
                fireDenied();
            else
                fireGranted();
            return;
        }
    }
/////////////////////////////////////////////////////////////////////////////////////////////////////////////

    private void fireGranted() {
        PermissionObserver.getInstance().notifyObservers();
        finish();
        overridePendingTransition(0, 0);
    }

    private void fireDenied() {
        final List<String> deniedPermissions = PermissionRequest.getDeniedPermissions(mContext, mRequestedPermissions);
        PermissionObserver.getInstance().notifyObservers(deniedPermissions);
        finish();
        overridePendingTransition(0, 0);
    }

    /////////////////////////////////////////////////////////////////////////////////////////////////////////////
    @Override
    public void setRequestedOrientation(int requestedOrientation) {
        try {
            super.setRequestedOrientation(requestedOrientation);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: permission/src/main/java/dev/eastar/permission/PermissionChecker.kt
/*
 * Copyright 2016 copyright eastar Jeong
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dev.eastar.permission

import android.content.ActivityNotFoundException
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.provider.Settings
import android.view.WindowManager
import androidx.appcompat.app.AlertDialog
import androidx.core.app.ActivityCompat

class PermissionChecker : androidx.appcompat.app.AppCompatActivity() {
    private var mRequestedPermissions: List<String> = emptyList()//최초 물어본 권한
    private var mRequestMessage: CharSequence? = null
    private var mDenyMessage: CharSequence? = null
    private var mRequestPositiveButtonText: CharSequence? = null
    private var mRequestNegativeButtonText: CharSequence? = null
    private var mDenyPositiveButtonText: CharSequence? = null
    private var mDenyNegativeButtonText: CharSequence? = null

    private var dlg: AlertDialog? = null

    interface EXTRA {
        companion object {
            const val PERMISSIONS = "PERMISSIONS"

            const val REQUEST_MESSAGE = "REQUEST_MESSAGE"
            const val REQUEST_POSITIVE_BUTTON_TEXT = "REQUEST_POSITIVE_BUTTON_TEXT"
            const val REQUEST_NEGATIVE_BUTTON_TEXT = "REQUEST_NEGATIVE_BUTTON_TEXT"

            const val DENY_MESSAGE = "DENY_MESSAGE"
            const val DENY_POSITIVE_BUTTON_TEXT = "DENY_POSITIVE_BUTTON_TEXT"
            const val DENY_NEGATIVE_BUTTON_TEXT = "DENY_NEGATIVE_BUTTON_TEXT"
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        window.addFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE)
        parseExtra()
        load()
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        setIntent(intent)
        parseExtra()
        load()
    }

    override fun onDestroy() {
        super.onDestroy()
        dlg?.cancel()
        PermissionObserver.deleteObservers()
    }

    private fun parseExtra() {
        mRequestedPermissions = intent.getStringArrayListExtra(EXTRA.PERMISSIONS) ?: emptyList()

        mRequestMessage = intent.getCharSequenceExtra(EXTRA.REQUEST_MESSAGE)
        mRequestPositiveButtonText = intent.getCharSequenceExtra(EXTRA.REQUEST_POSITIVE_BUTTON_TEXT)
        mRequestNegativeButtonText = intent.getCharSequenceExtra(EXTRA.REQUEST_NEGATIVE_BUTTON_TEXT)

        mDenyMessage = intent.getCharSequenceExtra(EXTRA.DENY_MESSAGE)
        mDenyPositiveButtonText = intent.getCharSequenceExtra(EXTRA.DENY_POSITIVE_BUTTON_TEXT)
        mDenyNegativeButtonText = intent.getCharSequenceExtra(EXTRA.DENY_NEGATIVE_BUTTON_TEXT)
    }

    private fun load() {
        val deniedPermissions = PermissionRequest.getDeniedPermissions(this, mRequestedPermissions)
        if (deniedPermissions.isEmpty()) {
            fireGranted()
            return
        }
        requestPermissions(deniedPermissions)
    }

    private fun requestPermissions(deniedPermissions: List<String>) {
        val message = mRequestMessage
        val permissions = deniedPermissions.toTypedArray()

        if (message.isNullOrBlank()) {
            ActivityCompat.requestPermissions(this, permissions, REQ_REQUEST)
            return
        }

        val context = this
        if (mRequestPositiveButtonText.isNullOrBlank())
            mRequestPositiveButtonText = getString(R.string.permission_confirm)
        if (mRequestNegativeButtonText.isNullOrBlank())
            mRequestNegativeButtonText = getString(R.string.permission_close)
        dlg = AlertDialog.Builder(context)
                .setMessage(message)
                .setOnCancelListener { fireDenied() }
                .setPositiveButton(mRequestPositiveButtonText) { _, _ -> ActivityCompat.requestPermissions(this, deniedPermissions.toTypedArray(), REQ_REQUEST) }
                .setNegativeButton(mRequestNegativeButtonText) { _, _ -> fireDenied() }
                .show()
        dlg?.setCanceledOnTouchOutside(false)
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (requestCode == REQ_REQUEST) {
            val deniedPermissions = PermissionRequest.getDeniedPermissions(this, mRequestedPermissions)
            if (deniedPermissions.isNotEmpty())
                denyPermissions()
            else
                fireGranted()
        }
    }

    private fun denyPermissions() {
        val context = this
        val message = mDenyMessage

        if (message.isNullOrBlank()) {
            fireDenied()
            return
        }

        if (mDenyPositiveButtonText.isNullOrBlank())
            mDenyPositiveButtonText = getString(R.string.permission_close)
        if (mDenyNegativeButtonText.isNullOrBlank())
            mDenyNegativeButtonText = getString(R.string.permission_setting)

        dlg = AlertDialog.Builder(context)
                .setMessage(message)
                .setOnCancelListener { fireDenied() }
                .setPositiveButton(mDenyPositiveButtonText) { _, _ -> fireDenied() }
                .setNegativeButton(mDenyNegativeButtonText) { _, _ ->
                    try {
                        startActivityForResult(Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS).setData(Uri.parse("package:" + context.packageName)), REQ_SETTING)
                    } catch (e: ActivityNotFoundException) {
                        startActivityForResult(Intent(Settings.ACTION_MANAGE_APPLICATIONS_SETTINGS), REQ_SETTING)
                    }
                }
                .show()
        dlg?.setCanceledOnTouchOutside(false)
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQ_SETTING) {
            val deniedPermissions = PermissionRequest.getDeniedPermissions(this, mRequestedPermissions)
            if (deniedPermissions.isNotEmpty())
                fireDenied()
            else
                fireGranted()
            return
        }
    }
    /////////////////////////////////////////////////////////////////////////////////////////////////////////////

    private fun fireGranted() {
        PermissionObserver.notifyObservers()
        finish()
        overridePendingTransition(0, 0)
    }

    private fun fireDenied() {
        val deniedPermissions = PermissionRequest.getDeniedPermissions(this, mRequestedPermissions)
        PermissionObserver.notifyObservers(deniedPermissions)
        finish()
        overridePendingTransition(0, 0)
    }

    /////////////////////////////////////////////////////////////////////////////////////////////////////////////
    override fun setRequestedOrientation(requestedOrientation: Int) {
        try {
            super.setRequestedOrientation(requestedOrientation)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    companion object {
        private const val REQ_REQUEST = 10
        private const val REQ_SETTING = 20
    }
}
